spec/03-plugins/38-ai-proxy/ai-proxy/00-config_spec.lua

```diff
@@ -0,0 +1,305 @@
+local PLUGIN_NAME = "ai-proxy"
+
+
+-- helper function to validate data against a schema
+local validate do
+  local validate_entity = require("spec.helpers").validate_plugin_config_schema
+  local plugin_schema = require("kong.plugins."..PLUGIN_NAME..".schema")
+
+  function validate(data)
+    return validate_entity(data, plugin_schema)
+  end
+end
+
+local WWW_MODELS = {
+  "openai",
+  "azure",
+  "anthropic",
+  "cohere",
+}
+
+local SELF_HOSTED_MODELS = {
+  "mistral",
+  "llama2",
+}
+
+describe(PLUGIN_NAME .. ": (schema)", function()
+
+
+  for i, v in ipairs(SELF_HOSTED_MODELS) do
+    it("requires upstream_url when using self-hosted " .. v .. " model", function()
+      local config = {
+        route_type = "llm/v1/chat",
+        auth = {
+          header_name = "Authorization",
+          header_value = "Bearer token",
+        },
+        model = {
+          name = "llama-2-7b-chat-hf",
+          provider = v,
+          options = {
+            max_tokens = 256,
+            temperature = 1.0,
+          },
+        },
+      }
+
+      if v == "llama2" then
+        config.model.options.llama2_format = "raw"
+      end
+
+      local ok, err = validate(config)
+
+      assert.not_nil(err["@entity"])
+      assert.not_nil(err["@entity"][1])
+      assert.equal(err["@entity"][1], "[model.options.upstream_url] must be specified when using the " .. v .. " provider")
```

Comment:
Ignore me if this doesn't apply to your tests, but there is a very helpful assertion called `same()` which does a deep comparison on inputs. Sometimes you can replace a whole bunch of `assert.equal()`'s with one `assert.same()` and get simpler test code:

```suggestion
      assert.same({
        "[model.options.upstream_url] must be specified when using the " .. v .. " provider",
      }, err["@entity"])
```

There are also `is_<type>` assertions that will produce better failure messages:

```lua
assert.is_table({}, "expected a table!")
assert.is_number(123)
```

Reply:
No this is all deliberate, as I don't want to test the entire pretend response - there are too many fields.

Just need to know the core info is correct.